core/tests/transport_upgrade.rs

```diff
@@ -118,12 +119,17 @@ fn upgrade_pipeline() {
     let listen_addr1 = Multiaddr::from(Protocol::Memory(random::<u64>()));
     let listen_addr2 = listen_addr1.clone();
 
-    let mut listener = listener_transport.listen_on(listen_addr1).unwrap();
+    listener_transport
+        .listen_on(ListenerId::new(1), listen_addr1)
+        .unwrap();
 
     let server = async move {
         loop {
-            let (upgrade, _remote_addr) =
-                match listener.next().await.unwrap().unwrap().into_upgrade() {
+            let (upgrade, _send_back_addr) =
+                match poll_fn(|cx| Pin::new(&mut listener_transport).poll(cx))
+                    .await
```

Comment:
Polling transports this way is rather inconvenient. Should we auto-implement `Stream` for transports?

Reply:
I'd be in favor yes.

Reply:
`impl<T: Transport> Stream for T` sounds good to me.

Reply:
> impl<T: Transport> Stream for T sounds good to me

Rust unfortunately does not allow this (I forget that every single time). 
I considered adding a `StreamTransport` that just wraps a transport and implemented `Stream`, but since we already have the `Boxed` type I just implemented it for `Boxed` and adjusted the tests (bfd5fb0). Also implemented Stream for `GenTcpTransport` (90721b9) because you can't box a TCP transport directly.